[Metrics Generator] Add support for entity limiting

Logiraptor opened the pull request:
This PR adds a new type of limiter to the metrics generator. It is
designed to limit on "entities" rather than series. In this case an
entity is a single labelset across multiple metrics, excluding external
labels. In effect, it allows the limiter to always generate the full set
of data for a given entity, rather than limiting randomly once the
series limit is triggered.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/grafana/tempo?pullRequest=5788) <br/>All committers have signed the CLA.

joe-elliott commented on modules/generator/registry/registry.go:
```diff
@@ -208,20 +240,48 @@ func (r *ManagedRegistry) onAddMetricSeries(count uint32) bool {
+		r.limitLogger.Log("msg", "max active entities reached", "rejected", len(rejected), "updated_since_last_collection", len(hashes))
+	}
+
+	rejectedMap := make(map[uint64]struct{})
```
i'm still digging on what this deleteFunc does but why not have TrackEntities return a `map[uint64]struct{}` to save allocs?

joe-elliott commented on modules/generator/registry/registry.go:
```diff
@@ -208,20 +240,48 @@ func (r *ManagedRegistry) onAddMetricSeries(count uint32) bool {
 	var seriesDemand int
 
 	for _, m := range r.metrics {
+		m.deleteFunc(func(hash uint64, _ int64) bool {
```
ok, I think I'm seeing. this is another way to cleanup metrics. we still have the time based cleanup below, but this is how you are enforcing the "max active entities" concept. on each collection loop you are removing all series whose "entity" is not in the kept group.

Logiraptor replied:
That's right, it could also be another function, but this felt more right to me. It's the same interface as `slices.DeleteFunc` and is the pattern we use in adaptive metrics' series tracking to manage garbage collection.

Logiraptor replied:
One alternative would be to keep `removeStaleSeries` and instead add another `delete(hash)` method. WDYT?

Logiraptor commented on modules/generator/registry/registry.go:
```diff
@@ -208,20 +240,48 @@ func (r *ManagedRegistry) onAddMetricSeries(count uint32) bool {
 func (r *ManagedRegistry) CollectMetrics(ctx context.Context) {
+	r.updatedSinceLastCollectionMtx.RLock()
+	hashes := maps.Clone(r.updatedSinceLastCollection)
+	r.updatedSinceLastCollection = make(map[uint64]struct{})
```
3aac0d562674f0e756e1277d46e5b7acf960ba8b

Logiraptor commented on modules/generator/registry/registry.go:
```diff
@@ -208,20 +240,48 @@ func (r *ManagedRegistry) onAddMetricSeries(count uint32) bool {
 
 func (r *ManagedRegistry) CollectMetrics(ctx context.Context) {
+	r.updatedSinceLastCollectionMtx.RLock()
+	hashes := maps.Clone(r.updatedSinceLastCollection)
```
I'm happy to do that way too. In practice the grpc call I am thinking off took around 20ms p99, so on a collection interval of 15s for example, it should still be fine.

Logiraptor commented:
@joe-elliott Based on our conversations in Grafana Labs internal slack, I've made the following changes:

* Simplified the Limiter interface to the minimum needed for _this PR_ only
* Extracted series limiting into an implementation of Limiter
* Added a configuration option to pick which limiter applies

I opted to keep the demand metrics for both limiters always on, this way it's possible to estimate what a good entity limit would be when migrating. I've also moved the series limiter metrics into local_series_limiter.go and added equivalents for entity limiting, so it should all feel familiar for operators when switching.

joe-elliott commented on modules/generator/instance.go:
```diff
@@ -109,12 +112,21 @@ type instance struct {
+	case LimiterTypeSeries:
+		limiter = localserieslimiter.New(overrides.MetricsGeneratorMaxActiveSeries, instanceID, limitLogger)
+	case LimiterTypeEntity:
+		limiter = localentitylimiter.New(overrides.MetricsGeneratorMaxActiveEntities, instanceID, limitLogger)
```
default case that returns an error?

joe-elliott commented on modules/generator/localentitylimiter/local_entity_limiter.go:
```diff
@@ -0,0 +1,125 @@
+
+type LocalEntityLimiter struct {
+	tenant            string
+	entityLastUpdated map[uint64]uint32
```
entityLastUpdated -> entityActiveSeries? or something similar?

joe-elliott commented on modules/generator/localentitylimiter/local_entity_limiter.go:
```diff
@@ -0,0 +1,125 @@
+	maxEntities := l.maxEntityFunc(l.tenant)
+	if maxEntities != 0 && uint32(len(l.entityLastUpdated))+1 > maxEntities {
+		l.limitLogger.Log("msg", "reached max active entities", "active_entities", len(l.entityLastUpdated), "max_active_entities", maxEntities)
+		l.metricTotalEntitiesLimited.Add(float64(1))
```
we could track series limited as well? i don't know if that's important to the way you want to operate this. up to you!

joe-elliott commented on modules/generator/localentitylimiter/local_entity_limiter.go:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	l.metricActiveEntities.Set(float64(len(l.entityLastUpdated)))
+	l.metricTotalEntitiesRemoved.Add(float64(seriesCount))
```
.Add(1) ?

joe-elliott commented on modules/generator/localentitylimiter/local_entity_limiter.go:
```diff
@@ -0,0 +1,125 @@
+	activeSeries, ok := l.entityLastUpdated[labelHash]
+	if ok {
+		activeSeries -= seriesCount
+		if activeSeries <= 0 {
```
activeSeries is a uint. compare to 0 only? we could also store series count as an int in case there are edge cases where activeSeries < 0.

joe-elliott reviewed (commented):
added some thoughts. i really like the reorg with the limiter interface and i think this will be np to merge. can we get a changelog entry?

also it might be nice to add a few lines to a doc somewhere that this mode exists and why. unfortunately i couldn't find a good place to add it. maybe just note it in the [metrics generator config](https://grafana.com/docs/tempo/v2.5.x/configuration/#metrics-generator)? a full explanation of the feature and all the details is likely a bit much (unless you are super excited to add all that), but just documenting its existence might be nice to see if we get community adoption.

Logiraptor commented:
@joe-elliott I believe I've implemented all your feedback. Regarding docs, I think for now the config docs are sufficient. Once I get some experience with this in practice, and also work on a few other improvements, it might be worth adding / updating a docs page specifically for managing span metrics cardinality, we'll see.

joe-elliott reviewed (approved):
Loki Grafana Tempo Mimir

looks like some minor conflicts to cleanup then we can merge?
